canopen/profiles/p402.py

```diff
@@ -1,39 +1,195 @@
+    OPEN_LOOP_SCALAR_MODE = -1
+    OPEN_LOOP_VECTOR_MODE = -2
 
-# status word 0x6041 bitmask and values in the list in the dictionary value
-POWER_STATES_402 = {
-    'NOT READY TO SWITCH ON': [0x4F, 0x00],
-    'SWITCH ON DISABLED'    : [0x4F, 0x40],
-    'READY TO SWITCH ON'    : [0x6F, 0x21],
-    'SWITCHED ON'           : [0x6F, 0x23],
-    'OPERATION ENABLED'     : [0x6F, 0x27],
-    'FAULT'                 : [0x4F, 0x08],
-    'FAULT REACTION ACTIVE' : [0x4F, 0x0F],
-    'QUICK STOP ACTIVE'     : [0x6F, 0x07]
-}
-
-# control word 0x6040
-POWER_STATE_COMMANDS = {
-    'SWITCH ON DISABLED'    : 0x80,
-    'DISABLE VOLTAGE'       : 0x04,
-    'READY TO SWITCH ON'    : 0x06,
-    'SWITCHED ON'           : 0x07,
-    'OPERATION ENABLED'     : 0x0F,
-    'QUICK STOP ACTIVE'     : 0x02
-}
-
-COMMAND_TO_POWER_STATE = {
-    0x80: 'SWITCH ON DISABLED',
-    0x06: 'READY TO SWITCH ON',
-    0x07: 'SWITCHED ON',
-    0x0F: 'OPERATION ENABLED',
-    0x02: 'QUICK STOP ACTIVE'
-}
-
-
-class Node402(RemoteNode):
+    CODE2NAME = {
+        PROFILED_POSITION           : 'PROFILED POSITION',
+        VELOCITY                    : 'VELOCITY',
+        PROFILED_VELOCITY           : 'PROFILED VELOCITY',
+        PROFILED_TORQUE             : 'PROFILED TORQUE',
+        HOMING                      : 'HOMING',
+        INTERPOLATED_POSITION       : 'INTERPOLATED POSITION',
+        CYCLIC_SYNCHRONOUS_POSITION : 'CYCLIC SYNCHRONOUS POSITION',
+        CYCLIC_SYNCHRONOUS_VELOCITY : 'CYCLIC SYNCHRONOUS VELOCITY',
+        CYCLIC_SYNCHRONOUS_TORQUE   : 'CYCLIC SYNCHRONOUS TORQUE',
+        OPEN_LOOP_SCALAR_MODE       : 'OPEN LOOP SCALAR MODE',
+        OPEN_LOOP_VECTOR_MODE       : 'OPEN LOOP VECTOR MODE'
+    }
+
+    NAME2CODE = {
+        'PROFILED POSITION'             : PROFILED_POSITION,
+        'VELOCITY'                      : VELOCITY,
+        'PROFILED VELOCITY'             : PROFILED_VELOCITY,
+        'PROFILED TORQUE'               : PROFILED_TORQUE,
+        'HOMING'                        : HOMING,
+        'INTERPOLATED POSITION'         : INTERPOLATED_POSITION,
+        'CYCLIC SYNCHRONOUS POSITION'   : CYCLIC_SYNCHRONOUS_POSITION,
+        'CYCLIC SYNCHRONOUS VELOCITY'   : CYCLIC_SYNCHRONOUS_VELOCITY,
+        'CYCLIC SYNCHRONOUS TORQUE'     : CYCLIC_SYNCHRONOUS_TORQUE,
+        'OPEN LOOP SCALAR MODE'         : OPEN_LOOP_SCALAR_MODE,
+        'OPEN LOOP VECTOR MODE'         : OPEN_LOOP_VECTOR_MODE
+    }
+
+    SUPPORTED = {
+        'PROFILED POSITION'           : 0x1,
+        'VELOCITY'                    : 0x2,
+        'PROFILED VELOCITY'           : 0x4,
+        'PROFILED TORQUE'             : 0x8,
+        'HOMING'                      : 0x20,
+        'INTERPOLATED POSITION'       : 0x40,
+        'CYCLIC SYNCHRONOUS POSITION' : 0x80,
+        'CYCLIC SYNCHRONOUS VELOCITY' : 0x100,
+        'CYCLIC SYNCHRONOUS TORQUE'   : 0x200,
+        'OPEN LOOP SCALAR MODE'       : 0x10000,
+        'OPEN LOOP VECTOR MODE'       : 0x20000
```

Comment:
I don't think that non-standard modes should be here. Every driver has their own stack of non standard nodes which aren't compatible with each other and this is only going to cause problems down the line. There should be a mechanism to extend the supported modes with vendor specific values.

Reply:
I also agree. At the moment I put the modes, but to enable the correct usage of this library we should only support generic modes, and all drive specific modes should be enable through extending the class.